Recast/Source/RecastArea.cpp

```diff
@@ -513,79 +513,113 @@ void rcMarkConvexPolyArea(rcContext* context, const float* verts, const int numV
+		const int vertIndexC = (vertIndex + 1) % numVerts;
+		const float* vertA = &verts[vertIndexA * 3];
+		const float* vertB = &verts[vertIndexB * 3];
+		const float* vertC = &verts[vertIndexC * 3];
+
+        // From A to B on the x/z plane
+		float ABdeltaX = vertB[0] - vertA[0];
+		float ABdeltaZ = vertB[2] - vertA[2];
+		float ABsqMag = ABdeltaX * ABdeltaX + ABdeltaZ * ABdeltaZ;
+		if (ABsqMag > EPSILON) // Is the squared magnitude of the distance between A and B greater than zero?
 		{
-			d0 = 1.0f / rcSqrt(d0);
-			dx0 *= d0;
-			dy0 *= d0;
+            // Normalize the delta vector
+            float inverseMag = 1.0f / rcSqrt(ABsqMag);
+            ABdeltaX *= inverseMag;
+            ABdeltaZ *= inverseMag;
 		}
-		float dx1 = vc[0] - vb[0];
-		float dy1 = vc[2] - vb[2];
-		float d1 = dx1 * dx1 + dy1 * dy1;
-		if (d1 > 1e-6f)
+
+        // From B to C on the x/z plane
+		float BCdeltaX = vertC[0] - vertB[0];
+		float BCdeltaZ = vertC[2] - vertB[2];
+		float BCsqMag = BCdeltaX * BCdeltaX + BCdeltaZ * BCdeltaZ;
+		if (BCsqMag > EPSILON) // Is the squared magnitude of the distance between B and C greater than zero?
 		{
-			d1 = 1.0f / rcSqrt(d1);
-			dx1 *= d1;
-			dy1 *= d1;
+            // Normalize the delta vector
+            float inverseMag = 1.0f / rcSqrt(BCsqMag);
+            BCdeltaX *= inverseMag;
+            BCdeltaZ *= inverseMag;
 		}
-		const float dlx0 = -dy0;
-		const float dly0 = dx0;
-		const float dlx1 = -dy1;
-		const float dly1 = dx1;
-		float cross = dx1 * dy0 - dx0 * dy1;
-		float dmx = (dlx0 + dlx1) * 0.5f;
-		float dmy = (dly0 + dly1) * 0.5f;
-		float dmr2 = dmx * dmx + dmy * dmy;
-		bool bevel = dmr2 * MITER_LIMIT * MITER_LIMIT < 1.0f;
-		if (dmr2 > 1e-6f)
+
+        // The y component of the cross product of the two delta vectors.
+        // The X and Z components of the cross product are both zero because the two
+        // delta vectors fall on the x/z plane.
+        float cross = BCdeltaX * ABdeltaZ - ABdeltaX * BCdeltaZ;
+
+        // CCW perpendicular vector to AB.  The segment normal.
+		const float ABperpX = -ABdeltaZ;
+		const float ABperpZ = ABdeltaX;
+
+        // CCW perpendicular vector to BC.  The segment normal.
+		const float BCperpX = -BCdeltaZ;
+		const float BCperpZ = BCdeltaX;
+
+        // Average the two segment normals to get the vertex normal for B.  This isn't normalized yet.
+		float BNormalX = (ABperpX + BCperpX) * 0.5f;
+		float BNormalZ = (ABperpZ + BCperpZ) * 0.5f;
+
+		float BNormalSqMag = BNormalX * BNormalX + BNormalZ * BNormalZ;
+
+        // If the magnitude of the normal average is less than about .69444, the corner is an acute enough angle
+        // that the result should be beveled.
+		bool bevel = BNormalSqMag * MITER_LIMIT * MITER_LIMIT < 1.0f;
+
+        // Now we normalize the BNormal vector
+		if (BNormalSqMag > EPSILON)
 		{
-			const float scale = 1.0f / dmr2;
-			dmx *= scale;
-			dmy *= scale;
+			const float inverseMag = 1.0f / rcSqrt(BNormalSqMag);
```

Comment:
This is not correct, and it's not normal, it is "middle" or "average" offset vector.

The the middle vector is calculated as if the _edges_ were offset and we would calculate intersection of neighbour edges at each vertex. That is, the sharper the corner, the further the offset is projected (thus the bevel limit too, or else it would shoot to sunset). If we normalize the offset, then the corners will get squashed.

Reply:
You can quickly verify that on paper with a 90° corner. preSegNorm=(-1,0), curSegNorm(0,1), should give you (-1,1), while normalized version would be (-0.707, 0.707).

Miter limit should be the same as commonly used in graphics (e.g. https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-miterlimit). If my math is right, 1.2 is about 17.5° corner.

Reply:
I tried to find source for the algo, but failed. I picked it up from some code long ago (could have been libart). Since the PostScript line miter limit comes so naturally out of it, maybe it comes from that lineage.

One "derivation" I did for myself at some point was that if you have segment normals `na` and `nb`, you calculate the bisector `na+nb`, and then project the bisector to `na`, you get the bisectors length `d` on `na`. Finally you find a scaling factor `1/d` so that the projected bisector length is 1 on `na`.

Reply:
Aaaaah this makes a ton of sense.  Thanks for clarifying!